### ISSUE TITLE
Relative traversing down the DOM

### ISSUE THREAD

**Issue by sriharshacv** [nathanvda/cocoon#213]
Hi 
I am new to open-source projects. So please forgive my naiveness.
Recently I was using nathanvda/cocoon for creating some really deep nested forms. I faced an issue while inserting new nested fields into parents/ grand-parents' siblings.
data-association-insertion-traversal  and data-association-insertion-node when used to traverse up the DOM tree can access only the ancestors at best when using the jquery method "closest". I found no possible way to reach the ancestor's other children.
So I suggest adding  association-insertion-traversal-down, association-insertion-node-down variables to enable downward traversing from ancestors. I already over-rode the cocoon.js file in my application but I thought it would be nice if everyone has such traversal option.

Any suggestions?
Regards
Sriharsha

**Comment by bryan-ash** [comment]
I had to convert that example into HAML before I could see the nesting:

```HAML
.grandfather
  .aunt
    This is a repeating trend
  .mother
    %a.start
      This is a repeating trend
.grandfather
  .aunt
    I want this
  .mother
    %a.start
      I am here. This is 'link to add association class'
```

I think this is the same behavior I'm asking about on [stackoverflow](http://stackoverflow.com/questions/27747913/cocoon-jquery-append-items-to-relavent-table)

@precluder can you fork the project and commit your changes?  I'd like to see how you solved it.

**Comment by nathanvda** [comment]
Yep, :+1: for converting it to HAML.

I will check out your question on stackoverflow.

**Comment by precluder** [comment]
@bryan-ash
Thanks for asking. I just created a fork https://github.com/precluder/cocoon
In the options add 'data-association-insertion-traversal-down' and 'data-association-insertion-node-down' with your traversal method and downward selector respectively. It should solve the issue. Please let me know in case of any issue.
@nathanvda 
I shall submit a pull request in the mean time.

**Comment by precluder** [comment]
Unfortunately I was unable to write the test cases in javascript. So please verify the functionality just in case.

**Comment by nathanvda** [comment]
Could this specific case not be solved by using: 

     'data-association-insertion-method'    => 'append' 
     'data-association-insertion-node'      => '.aunt'
     'data-association-insertion-traversal' => 'closest'

**Comment by bryan-ash** [comment]
I tried using traversal => 'closest' in my project with no joy.  

With a breakpoint on line 55 in cocoon.js, I see: 
```javascript
insertionNode=".panel-items"
insertionTraversal='closest'

$this => [<a class="btw...]
$this[insertionTraversal](insertionNode) => []
$this[insertionTraversal]() => []

$this['parent']() => [<div class="table-responsive">]
$this['parent']('table') => []
```

my for fields are:
```haml
.table-responsive
  %table.table.table-hover
    %tbody.panel-items
      = panel_form.simple_fields_for :panel_items do |panel_item_fields|
        = render 'panel_item_fields', f: panel_item_fields
  = link_to_add_association panel_form, :panel_items,
      'data-association-insertion-method' => 'append',
      'data-association-insertion-node' => '.panel-items',
      'data-association-insertion-traversal' => 'closest',
      class: 'btn btn-success pull-right' do
    %span{class: 'glyphicon glyphicon-plus-sign'}
    New item
```

So I can navigate up to the parent div, but not down again.

**Comment by nathanvda** [comment]
My layouts are generally much simpler. I will not ask why you are using a table for layout when using bootstrap.

My first idea would be to place the `link_to_add_association` where it would work, and then add another link where you would want it, and when clicking the cosmetically correct link, trigger the click action for the link where the insert works. A bit of a workaround, and not even sure if it works. 

I am not too happy with the pull request by @precluder but it did give me an idea how to solve this better. I am going to try that out and let you know ...

**Comment by bryan-ash** [comment]
I agree that adding more data attributes is starting to get messy, and is limited to one-up-one-down anyway.  

A hook that allowed me to replace the javascript insertion function would be great.

i.e. if `insertionNode[insertionMethod](contentNode);` was in a function that I could override.

**Comment by jamesvanklink** [comment]
I know this is a bit old but still open so I thought I'd share my solution.  I had a similar situation and solved it by writing a custom jQuery function and using it as the traversal:

 $.fn.findAdderContainer = function(selector) {
    return this.parent().parent().children(selector);
  };

You can then just use the data attributes for cocoon as you normally would and this will return what you need.

**Comment by nathanvda** [comment]
Ow @jamesvanklink that looks awesome! I am going to try that out. Thanks :+1:

**Comment by nathanvda** [comment]
Nopw solved by #355 that allows the `data-association-insertion-node` to be a function, so you have all the freedom in the world to travers the DOM as you please.
